src/tactic/derive_fintype.lean

```diff
@@ -0,0 +1,242 @@
+/-
+Copyright (c) 2020 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+-/
+import logic.basic data.fintype.basic
+
+/-!
+# Derive handler for `fintype` instances
+
+This file introduces a derive handle to automatically generate `fintype`
+instances for structures and inductives.
+-/
+
+namespace derive_fintype
+
+/-- A step in the construction of `finset.univ` for a finite inductive type.
+We will set `enum` to the discriminant of the inductive type, so a `finset_above`
+represents a finset that enumerates all elements in a tail of the constructor list. -/
+def finset_above (α) (enum : α → ℕ) (n : ℕ) :=
+{s : finset α // ∀ x ∈ s, n ≤ enum x}
+
+/-- Construct a fintype instance from a completed `finset_above`. -/
+def mk_fintype {α} (enum : α → ℕ) (s : finset_above α enum 0) (H : ∀ x, x ∈ s.1) :
+  fintype α := ⟨s.1, H⟩
+
+/-- This is the case for a simple variant (no arguments) in an inductive type. -/
+def finset_above.cons {α} {enum : α → ℕ} (n)
+  (a : α) (h : enum a = n) (s : finset_above α enum (n+1)) : finset_above α enum n :=
+begin
+  refine ⟨finset.cons a s.1 _, _⟩,
+  { intro h',
+    have := s.2 _ h', rw h at this,
+    exact nat.not_succ_le_self n this },
+  { intros x h', rcases finset.mem_cons.1 h' with rfl | h',
+    { exact ge_of_eq h },
+    { exact nat.le_of_succ_le (s.2 _ h') } }
+end
+
+theorem finset_above.mem_cons_self {α} {enum : α → ℕ} {n a h s} :
+  a ∈ (@finset_above.cons α enum n a h s).1 := multiset.mem_cons_self _ _
+
+theorem finset_above.mem_cons_of_mem {α} {enum : α → ℕ} {n a h s b} :
+  b ∈ (s : finset_above _ _ _).1 → b ∈ (@finset_above.cons α enum n a h s).1 :=
+multiset.mem_cons_of_mem
+
+/-- The base case is when we run out of variants; we just put an empty finset at the end. -/
+def finset_above.nil {α} {enum : α → ℕ} (n) : finset_above α enum n := ⟨∅, by rintro _ ⟨⟩⟩
+
+/-- This is a finset covering a nontrivial variant (with one or more constructor arguments).
+The property `P` here is `λ a, enum a = n` where `n` is the discriminant for the current
+variant. -/
+def finset_in {α} (P : α → Prop) := {s : finset α // ∀ x ∈ s, P x}
+
+/-- To construct the finset, we use an injective map from the type `Γ`, which will be the
+sigma over all constructor arguments. We use sigma instances and existing fintype instances
+to prove that `Γ` is a fintype, and construct the function `f` that maps `⟨a, b, c, ...⟩`
+to `C_n a b c ...` where `C_n` is the nth constructor, and `mem` asserts
+`enum (C_n a b c ...) = n`. -/
+def finset_in.mk {α} {P : α → Prop} (Γ) [fintype Γ]
```

Comment:
Some variables here could be implicit, but I see you're already using @ below, so maybe the choice is intentional? Same in `finset_above.union`, `n` can be implicit.

Reply:
The implicitness here is deliberate for the use case in the tactic. In the case of `n` in `finset_above.union` and `finset_above.cons`, I explicitly insert numerals at each stage so that it doesn't build up a `0+1+1` numeral, and here the `Γ` is explicit because it is the first thing to be supplied, in a tactic invocation like `apply finset_in.mk Γ` (letting the instance search handle the fintype proof), and `f` and other things are proven as subgoals.